### PR TITLE
Add AsyncPage implementation and docs. Add related tests

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds docs to the `AsyncPage` interface, adds the `AsyncPageImpl` implementation and related tests. This adds support for asynchronously fetching paginated results.

### CODE REVIEW

File: gcloud-java-core/src/main/java/com/google/cloud/AsyncPageImpl.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Base implementation for asynchronously consuming Google Cloud paginated results.
+ *
+ * @param <T> the value type that the page holds
+ */
+public class AsyncPageImpl<T> extends PageImpl<T> implements AsyncPage<T> {
+
+  private static final long serialVersionUID = -6009473188630364906L;
+
+  private NextPageFetcher<T> asyncPageFetcher;
```
Comment (mziccard) [inline gcloud-java-core/src/main/java/com/google/cloud/AsyncPageImpl.java]: Codacy Issue found: [Private field 'asyncPageFetcher' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2499477474/issues/source?bid=3123109&fileBranchId=3271690#l35)

File: gcloud-java-core/src/main/java/com/google/cloud/AsyncPageImpl.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  public interface NextPageFetcher<T> extends Serializable {
+    Future<AsyncPage<T>> nextPage();
+  }
+
+  private static class SyncNextPageFetcher<T> implements PageImpl.NextPageFetcher<T> {
+
+    private static final long serialVersionUID = -4124568632363525351L;
+
+    private NextPageFetcher<T> asyncPageFetcher;
```
Comment (mziccard) [inline gcloud-java-core/src/main/java/com/google/cloud/AsyncPageImpl.java]: Codacy Issue found: [Private field 'asyncPageFetcher' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2509268387/issues/source?bid=3123109&fileBranchId=3271690#l50)

File: gcloud-java-core/src/main/java/com/google/cloud/AsyncPage.java [hunk 3]
```diff
@@ -18,6 +18,34 @@
...
+ * }}</pre>
+ *
+ * <p>Or handle pagination explicitly:
+ * <pre> {@code
+ * AsyncPage<T> page = ...; // get a AsyncPage<T> instance
+ * while (page != null) {
+ *   for (T value : page.values()) {
+ *     // do something with value
+ *   }
+ *   page = page.nextPage().get();
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/cloud/AsyncPage.java]: shouldn't that be `page.nextPageAsync().get();` ?

Comment (mziccard) [inline gcloud-java-core/src/main/java/com/google/cloud/AsyncPage.java]: Yes totally.

**Pull request merged**